benches/dapol.rs

```diff
@@ -1,175 +1,350 @@
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
+    let num_entities = NUM_USERS[2]; // 30_000: max. value for tree height 16
+    let thread_counts: [u8; 7] = [4, 8, 16, 32, 64, 128, 255];
+
+    for h in TREE_HEIGHTS {
+        let ndm_smt =
+            setup::build_ndm_smt(Height::from(h), MaxThreadCount::default(), num_entities);
+        let entity_id = EntityId::from_str("foo").unwrap();
+        let proof = setup::generate_proof(&ndm_smt, &entity_id);
+
+        group.bench_function(BenchmarkId::new("tree_height", h), |bench| {
+            bench.iter(|| {
+                proof
+                    .verify(ndm_smt.root_hash())
+                    .expect("Unable to verify proof")
+            })
         });
+    }
+
+    for t in thread_counts {
+        let ndm_smt = setup::build_ndm_smt(Height::from(16), MaxThreadCount::from(t), num_entities);
+        let entity_id = EntityId::from_str("foo").unwrap();
+        let proof = setup::generate_proof(&ndm_smt, &entity_id);
 
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("padding", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
+        group.bench_function(BenchmarkId::new("max_thread_count", t), |bench| {
+            bench.iter(|| {
+                proof
+                    .verify(ndm_smt.root_hash())
+                    .expect("Unable to verify proof")
+            })
+        });
+    }
+
+    for u in NUM_USERS {
+        let ndm_smt = setup::build_ndm_smt(Height::from(16), MaxThreadCount::default(), u);
+        let entity_id = EntityId::from_str("foo").unwrap();
+        let proof = setup::generate_proof(&ndm_smt, &entity_id);
+
+        group.bench_function(BenchmarkId::new("num_users", u), |bench| {
+            bench.iter(|| {
+                proof
+                    .verify(ndm_smt.root_hash())
+                    .expect("Unable to verify proof")
+            })
         });
     }
 
     group.finish();
 }
 
-criterion_group!(dapol_group, build_dapol, generate_proof, verify_proof);
-criterion_main!(dapol_group);
-
-// HELPER FUNCTIONS
+// BENCHMARKS: IAI
```

Comment:
If you do decide to use IAI for mem measurements as apposed to custom criterion measurement then can you move this code to a new PR. Also, since it's such a pain to write all the functions just skip a bunch of them. Like only do a fraction of the full vector space.

Reply:
Will do